server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/message/view/MessageFastViewFactory.java

```diff
@@ -0,0 +1,169 @@
+                 return messageFullViewFactory.mainTextContent(messageContent)
+                    .map(Preview::compute);
+            } catch (MailboxException e) {
+                LOGGER.error("cannot extract content of the message {}", mimeMessage.getMessageId());
+                return Optional.empty();
+            }
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MessageFastViewFactory.class);
+
+    private final BlobManager blobManager;
+    private final MessageIdManager messageIdManager;
+    private final MessageFastViewProjection fastViewProjection;
+    private final MessageFullViewFactory messageFullViewFactory;
+
+    @Inject
+    @VisibleForTesting
+    MessageFastViewFactory(BlobManager blobManager, MessageIdManager messageIdManager, MessageFastViewProjection fastViewProjection,
+                           MessageFullViewFactory messageFullViewFactory) {
+        this.blobManager = blobManager;
+        this.messageIdManager = messageIdManager;
+        this.fastViewProjection = fastViewProjection;
+        this.messageFullViewFactory = messageFullViewFactory;
+    }
+
+    @Override
+    public List<MessageFastView> fromMessageIds(List<MessageId> messageIds, MailboxSession mailboxSession) {
+        return Mono.from(fastViewProjection.retrieve(messageIds))
+            .flatMap(previews -> gatherMessageViews(messageIds, mailboxSession, previews))
+            .subscribeOn(Schedulers.boundedElastic())
+            .block();
+    }
+
+    private Mono<List<MessageFastView>> gatherMessageViews(List<MessageId> messageIds, MailboxSession mailboxSession,
+                                                           Map<MessageId, MessageFastViewPrecomputedProperties> fastProjections) {
+        return Flux.merge(
+                fetch(ImmutableList.copyOf(fastProjections.keySet()), FetchGroup.HEADERS, mailboxSession),
+                fetch(withoutPreviews(messageIds, fastProjections), FetchGroup.FULL_CONTENT, mailboxSession))
+            .collectList()
+            .map(messages -> Helpers.toMessageViews(messages, new FromMessageResultAndPreview(blobManager, messageFullViewFactory, fastProjections)));
+    }
+
+    private List<MessageId> withoutPreviews(List<MessageId> messageIds, Map<MessageId, MessageFastViewPrecomputedProperties> fastProjections) {
+        return messageIds.stream()
+            .filter(id -> !fastProjections.containsKey(id))
+            .collect(Guavate.toImmutableList());
+    }
+
+    private Flux<MessageResult> fetch(List<MessageId> messageIds, FetchGroup fetchGroup, MailboxSession mailboxSession) {
+        return Mono.fromCallable(() -> messageIdManager.getMessages(messageIds, fetchGroup, mailboxSession))
+            .onErrorResume(MailboxException.class, ex -> {
+                LOGGER.error("cannot read messages {}", messageIds, ex);
+                return Mono.just(ImmutableList.of());
+            })
+            .flatMapMany(Flux::fromIterable);
```

Comment:
Write this with Flux.fromIterable & a try/catch + flux.empty, I believe it will be more readable.

Reply:
I can't, putting the operation outside of callbacks make it executed in a blocking way